cli/spec.py

```diff
@@ -0,0 +1,244 @@
+import ConfigParser
+import os
+import sys
+import yaml
+
+import clg
+import yamlordereddictloader
+from cli import exceptions
+from cli import utils
+
+SPEC_EXTENSION = '.spec'
+LOOKUP_DICT = {
+    '__DEFAULT__': "default"
+}
+
+TRIM_PARAMS = ['default', 'required', 'requires_only']
+
+
+def clg_ini_file(value):
+    """
+    The custom type for clg spec
+    :param value: the argument value.
+    :return: dict based on a provided ini file.
+    """
+    _config = ConfigParser.ConfigParser()
+    _config.read(value)
+
+    d = dict(_config._sections)
+    for k in d:
+        d[k] = dict(_config._defaults, **d[k])
+        for key, value in d[k].iteritems():
+            # check if we have lists
+            value = utils.string_to_list(value, append_to_list=False)
+            d[k][key] = value
+        d[k].pop('__name__', None)
+
+    return d
+
+
+def parse_args(module_name, config, args=None):
+    """
+    Looks for all the specs for specified module
+    and parses the commandline input arguments accordingly.
+
+    :param module_name: the module name: installer|provisioner|tester
+    :param config: the infrared configuration file
+    :param args: additional arguments to pass to the clg.
+    """
+
+    clg_options, global_opts, subcommand_opts = \
+        __load_spec_as_dict(module_name, config)
+
+    cmd = clg.CommandLine(clg_options)
+    res_args = vars(cmd.parse(args))
+
+    # get the default values
+    command_args = subcommand_opts[res_args['command0']]
+    defaults = {k: v.get('default', None)
+                for k, v in command_args.iteritems()}
+
+    # generate config file if required
+    if res_args['generate-conf-file']:
+        try:
+            file_name = res_args['generate-conf-file']
+            out_config = ConfigParser.ConfigParser()
+
+            section = res_args['command0']
+            out_config.add_section(section)
+            for opt, value in defaults.iteritems():
+                if value is not None:
+                    out_config.set(section, opt, value)
+            with open(file_name, 'w') as configfile:    # save
+                out_config.write(configfile)
+        except Exception as ex:
+            raise exceptions.IRException(ex.message)
+        finally:
+            sys.exit(0)
```

Comment:
Two things in here:
1. Try to eliminate of calling sys.exit (especially outside the main method)
2. What will happen when exception is raised?!?